core/src/main/java/io/questdb/std/DecimalKnuthDivider.java

```diff
@@ -0,0 +1,468 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.std;
+
+import java.math.RoundingMode;
+import java.util.Arrays;
+
+/**
+ * Thread-safe implementation of Knuth's division algorithm (Algorithm D from TAOCP Vol 2, Section 4.3.1).
+ * This class provides high-precision division for 256-bit decimal numbers using the classical
+ * long division algorithm optimized for multi-precision arithmetic.
+ * <p>
+ * The algorithm divides a dividend u by a divisor v, producing quotient q and remainder r
+ * such that u = q * v + r, where 0 <= r < v.
+ */
+class DecimalKnuthDivider {
+    private static final ThreadLocal<DecimalKnuthDivider> tl = new ThreadLocal<>(DecimalKnuthDivider::new);
+    private final int[] q = new int[8];
+    private final int[] u = new int[9];
+    private final int[] v = new int[8];
+    private int m = 0;
+    private int n = 0;
+
+    public static DecimalKnuthDivider instance() {
+        DecimalKnuthDivider instance = tl.get();
+        instance.clear();
+        return instance;
+    }
+
+    /**
+     * Performs the division operation and round depending on the remainder value and the isNegative.
+     * Note: if isNegative is true, the quotient won't reflect that, you will need to negate the result.
+     */
+    public void divide(boolean isNegative, RoundingMode roundingMode) {
+        // If the dividend is smaller than the divisor, the result is zero.
+        if (m < n) {
+            round(isNegative, roundingMode);
+            return;
+        }
+
+        // If the divisor is a word, we can use the optimized algorithm.
+        if (n == 1) {
+            divideByWord();
```

Comment:
it seems that most significant number after divideByWord() is 0. However, subsequent round() calculates this again wasting time